### PR TITLE
route53 provider and tests

### PR DESCRIPTION
**Pull request by eadmundo**
AWS Route 53 provider for lexicon

Notes:

* haven't (yet?) added any command line args for auth - I always use env vars or credentials files, and the provider uses boto under the hood, so follows the same [credential lookup approach](http://boto3.readthedocs.io/en/latest/guide/configuration.html). Maybe we should have `--aws-access-key-id` etc, interested in other thoughts.
* skipped one of the tests since AWS record sets don't have IDs
* to create the vcr cassettes, I ran the tests using my own domain & AWS keys, and then did search & replace in the cassette files for the domain
* if one deletes the hosted zone (domain equivalent) in Route 53 within 12 hours, there's no charge, enabling free testing (in theory)

### CODE REVIEW

**Comment by CapsuleCD** [comment]
Hi.

I'm an automated pull request bot named [CapsuleCD](http://www.github.com/AnalogJ/capsulecd). I handle testing, versioning and package releases for this project. 
- If you're the owner of this repo, you can click the button below to kick off the CapsuleCD build pipeline and create an automated release.'
- If not, don't worry, someone will be by shortly to check on your pull request. 

[CapsuleCD](http://www.capsulecd.com/#!/deploy/github/AnalogJ/lexicon/pullrequests/57)

---
If you're interested in learning more about [CapsuleCD](http://www.github.com/AnalogJ/capsulecd), or adding it to your project, you can check it out [here](http://www.github.com/AnalogJ/capsulecd)

**Comment by eadmundo** [comment]
Also tests will need AWS credentials with [appropriate Route 53 access permissions](http://docs.aws.amazon.com/Route53/latest/DeveloperGuide/r53-api-permissions-ref.html#required-permissions-resource-record-sets)

**Comment by eadmundo** [comment]
boto is throwing a no credentials error on the tests - the simplest fix would be to put dummy keys into Circle CI.

**Comment by AnalogJ** [comment]
Hey @eadmundo Thanks for the PR. I'll take a look and merge it in tonight/tomorrow

**Comment by eadmundo** [comment]
@AnalogJ awesome, cheers.

File: tests/providers/integration_tests.py [hunk 1]
```diff
@@ -209,7 +209,8 @@ def test_Provider_when_calling_update_record_should_modify_record(self):
             provider.authenticate()
             assert provider.create_record('TXT','orig.test','challengetoken')
             records = provider.list_records('TXT','orig.test')
-            assert provider.update_record(records[0]['id'],'TXT','updated.test','challengetoken')
+            print records[0]
```
Comment (AnalogJ) [inline tests/providers/integration_tests.py]: please remote debug statements from test suite.

File: tests/providers/integration_tests.py [hunk 2]
```diff
@@ -196,9 +196,9 @@ def test_Provider_when_calling_list_records_with_arguments_should_filter_list(se
             provider.authenticate()
             assert isinstance(provider.list_records(), list)
 
-    ###########################################################################
-    # Provider.update_record()
-    ###########################################################################
+    # ###########################################################################
```
Comment (AnalogJ) [inline tests/providers/integration_tests.py]: not sure why this is different.

File: tests/providers/integration_tests.py [hunk 3]
```diff
@@ -233,12 +234,14 @@ def test_Provider_when_calling_update_record_with_fqdn_name_should_modify_record
...
+            assert provider.update_record(records[0].get('id', None),'TXT','updated.testfqdn.{0}.'.format(self.domain),'challengetoken')
 
-    ###########################################################################
-    # Provider.delete_record()
-    ###########################################################################
+    # ###########################################################################
+    # # Provider.delete_record()
+    # ###########################################################################
     def test_Provider_when_calling_delete_record_by_identifier_should_remove_record(self):
+        if self.provider_name in ['route53']:
```
Comment (AnalogJ) [inline tests/providers/integration_tests.py]: don't skip tests for a specific provider in the test. That can be done in the provider test wrapper. See PointHQ: https://github.com/AnalogJ/lexicon/blob/master/tests/providers/test_pointhq.py#L18

**Comment by AnalogJ** [comment]
Hey  @eadmundo I've added some inline comments to your PR, but here's some additional feedback:

- please don't dont add a dependency to requirements.txt. I don't want users who don't need Route53 validation to have to install boto. Instead add boto as a extra/optional dependency in `setup.py`. See [here for an example](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies). The tests will fail but I'll fix that. 
- I would add `Authorization` as a sensitive header for your recordings, so that it doesn't get recorded. See PointHQ for an example: https://github.com/AnalogJ/lexicon/blob/master/tests/providers/test_pointhq.py#L16
- Don't skip tests for a specific provider in the test. That can be done in the provider test wrapper. See PointHQ: https://github.com/AnalogJ/lexicon/blob/master/tests/providers/test_pointhq.py#L18-L19


If you can get those small issues fixed I should be able to merge the PR :)

Thanks again for all your work!

File: tests/providers/integration_tests.py [hunk 4]
```diff
@@ -102,9 +102,9 @@ def test_Provider_when_calling_create_record_for_TXT_with_fqdn_name_and_content(
             provider.authenticate()
             assert provider.create_record('TXT',"_acme-challenge.fqdn.{0}.".format(self.domain),'challengetoken')
 
-    ###########################################################################
-    # Provider.list_records()
-    ###########################################################################
+    # ###########################################################################
```
Comment (eadmundo) [inline tests/providers/integration_tests.py]: same

**Comment by eadmundo** [comment]
All sounds good, will fix up. Happy to contribute - lexicon is a fantastic tool!

**Comment by AnalogJ** [comment]
Hey @eadmundo I merged your PR and made some tweaks. Lexicon has a standardized way of passing credentials (either via the cmd line or using env variables). I updated your provider to follow that pattern (which unfortunately means it won't work out of the box with a ~/.aws/credentials file) 

I ended up adding `--auth-access-key` and `--auth-access-secret` flags to the CLI, but because of some legacy test code I also had to add `--auth-username` and `--auth-token` which I'll have to fix later. 

Thanks for all your help though!

**Pull request merged**